Moving code from gax-grpc to gax

garrettjonesgoogle opened the pull request:
~~This is just a work-in-progress PR, so I am just looking for feedback on overall design.~~ I'd like validation on the following changes:

1. Overall, I tried to move logic up from grpc/httpjson to gax.rpc
2. Moved most of the implementation of `GrpcCallableFactory` to `CallableFactory` in gax.rpc
3. Created an `OperationApi` abstraction in gax.rpc, along with an `OperationSnapshot` abstraction
  3.1. Now that `OperationSnapshot` is used throughout gax.rpc, `OperationT` is removed as a type parameterization
  3.2. Most of the operation implementation is moved up to gax.rpc
4. Renamed `Transport` to `TransportChannel` and `ChannelProvider` to `TransportChannelProvider`; removed `TransportProvider`, and instead repurposed Instantiating(*)ChannelProvider to directly create `TransportChannel` objects (either `GrpcTransportChannel` or `HttpJsonTransportChannel`), cutting out a layer of abstraction
5. Created new `TransportDescriptor` concept, to allow for transport-specific logic to be supplied in various places
6. Moved header logic out of Instantiating(*)ChannelProvider into a new `HeaderProvider`, which lives at the gax.rpc level and can be shared by grpc & http/json
7. In the new `TransportChannelProvider`, changed the model of providing other providers, such that you can do provider.withX(X x), so now there is only a single `getTransportChannel()` method

pongad commented:
Can this be split at all? The individual bullets in the description sound reasonable, but I'm having great difficulty groking the 7000 line diff

vam-google commented on gax/src/main/java/com/google/api/gax/rpc/testing/FakeApiExceptionFactory.java:
```diff
@@ -0,0 +1,131 @@
+        return new DeadlineExceededException(cause, grpcStatusCode, retryable);
+      case ALREADY_EXISTS:
+        return new AlreadyExistsException(cause, grpcStatusCode, retryable);
+        //      case PERMISSION_DENIED:
```
What do all this commented string mean? I don't see any related FIXME or TODO.

garrettjonesgoogle replied:
It's because I copied the full set from the real exception factory and commented out the ones that I didn't define in the fake enum.

vam-google commented on gax/src/main/java/com/google/api/gax/rpc/testing/FakeStatusCode.java:
```diff
@@ -0,0 +1,79 @@
+/*
```
Why do we have so many `Fake*` classes in `main` (not `test`) directory? Are ther really necessary? Can't many of them be replaced by just mockito mocks? Can't they be in `test/` folder instead of `main/*/testing`?

garrettjonesgoogle replied:
I believe the original reason was so that the classes could be used across package boundaries, e.g. in gax-grpc - there is currently at least one thing being referenced. It's possible some of them could be replaced by mockito mocks, but I think it's actually nice to implement the classes this way as a validation at the core level of the implementability of the interfaces.

vam-google replied:
What about moving them in main/*/testing? They will still be accessible in tests across package boundaries. This is really ugly to have so much "test garbage" in production code/artifact.

vam-google commented on gax/src/test/java/com/google/api/gax/rpc/OperationCallableImplTest.java:
```diff
@@ -0,0 +1,983 @@
+/*
```
Why is this test marked as new not moved? It is important for this one, because it is a huge and important test, which test stuff, which was heavily modified as part of this PR. It is hard to review it porperly, when everything is green.

garrettjonesgoogle replied:
It's weird, it seems like the files should be close enough to match. Darn git! 

Maybe I'll create a "Diff PR" like I did for a previous large PR.

vam-google commented on gax/src/test/java/com/google/api/gax/rpc/OperationCallableImplTest.java:
```diff
@@ -0,0 +1,983 @@
+    initialContext = getClientContext(initialChannel, executor);
+  }
+
+  private static class ResponseTransformer implements ApiFunction<OperationSnapshot, Color> {
```
This is suspicious. Those transformers are essential part of what was tested by this test. Now, they are the fake ones, as result it looks like the test is not testing actual operations anymore (it is impoortant, because the longrunning operations logic relies on those transformers to be correct).

garrettjonesgoogle replied:
This is a consequence of the operation logic moving to the core level; the transformers are transport-specific and can't be moved up. We'll need to have tests at the grpc level for the transformations (still TO DO for the PR).

vam-google replied:
Are you going to add the tests as part of this PR?

vam-google commented on gax/src/test/java/com/google/api/gax/rpc/OperationCallableImplTest.java:
```diff
@@ -0,0 +1,983 @@
+      public ApiFuture<OperationSnapshot> futureCall(RequestT request, ApiCallContext context) {
+        OperationSnapshot response = results[index];
+        if (index < results.length - 1) {
+          index += 1;
```
`+= 1`? Why not `++`?

garrettjonesgoogle replied:
Personal preference: I like the uniformity of using `+=` for all values on the right hand side of the expression.

vam-google commented on gax/src/test/java/com/google/api/gax/rpc/OperationCallableImplTest.java:
```diff
@@ -0,0 +1,983 @@
+      case CANCELLED:
+        expectedClass = CancelledException.class;
+        break;
+        //      case NOT_FOUND:
```
Why are all these commented?

garrettjonesgoogle replied:
I copied the function from somewhere else, and commented out the enum values that I hadn't defined yet in `FakeStatusCode.Code`.

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcOperationApi.java:
```diff
@@ -0,0 +1,108 @@
+            return CancelOperationRequest.newBuilder().setName(request).build();
+          }
+        },
+        new ApiFunction<Empty, Void>() {
```
It seems like a no-op transformer. Is it possible inside `TranslatingUnaryCallable` to just not do any transformation in such cases (this allows to avoid doing heavy ApiFutures.transform() when it is not required).

garrettjonesgoogle replied:
Unfortunately you can't just return an `Empty` as a `Void` (the first will have a default instance, the second has to be `null`).

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcOperationApi.java:
```diff
@@ -0,0 +1,108 @@
+        new ApiFunction<Operation, OperationSnapshot>() {
+          @Override
+          public OperationSnapshot apply(Operation operation) {
+            return GrpcOperationSnapshot.create(operation);
```
So this allows to wrap Operation in OperationSnapshot, but I still don't understand what exactly problem is solved by OperationSnapshot? Allows to remvoe 4th parametrized parameter?

I still don't have good understanding on what is happening here (so sorry if the following does not make sense). It seems like `OperationApi + OperationSnapshot` are trying to build a manual-on-gapic client on `OperationsClient`, whithout actually doing one. Maybe it is better to wrap all that into a manual on top of gapic's `OperationsClient` (i.e. create some wraper around, which will return OperationSnapshot instead Operation). 

Also, TranslatingUnaryCallable comes with cost of ApiFutures.transform() on operation result. Those transforms are not as smooth as we would like them to be (they imply bidirectional (from wrapper to wrapped obj, and from wrapped obj to wrapper)) completion/failure/cancelation propagation. This is a very hard goal to achieve and it doesn't work under all circumstances (when there are unexpected errors in the middle, for example in transformers themselves). I would really like to avoid transforms if not absolutely necessary.

garrettjonesgoogle replied:
The creation of `OperationSnapshot` is a consequence of creating `OperationApi` in core - you need some way to return the `Operation` data back. If you use a parameterized type, then you can't perform any logic on it, e.g. the polling logic that I'm moving up to core. 

Yes, this is kind of a manual-on-gapic client, but it also has to be possible for the interface to be backed by a non-gapic data source - for http1 APIs, this may look very different. 

I don't fully understand your concerns with the transformers. They are really just type transformers on the input/output types, and I didn't think they would affect completion/failure/cancelation propagation.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/OperationApi.java:
```diff
@@ -30,36 +30,13 @@
-/** The settings used to create a Transport. */
 @BetaApi
-public interface TransportProvider {
+public interface OperationApi {
```
I will move operation stuff that is not rpc-specific (e.g. callable-related) to `com.google.api.gax.longrunning`.

Also, related to the `OperationApi` name, I will rename that to `LongRunningClient`. It does need to stay under `com.google.api.gax.rpc` though since it references `UnaryCallable`.

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcOperationSnapshot.java:
```diff
@@ -29,41 +29,45 @@
-@BetaApi
-public final class FixedChannelProvider implements ChannelProvider {
-  private final ManagedChannel channel;
+public class GrpcOperationSnapshot implements OperationSnapshot {
```
This is an abstraction of the Operation proto-generated type, return returns an `Any` for `getMetadata()` and `getResponse()`. The typed data is surfaced by `OperationCallable` / `OperationFuture` which use the response / metadata transformers to convert this `Object` to the type expected by the client. 

Basically I think this is lower level than you were expecting.

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcStatusCodes.java:
```diff
@@ -0,0 +1,50 @@
+
+  public static Set<Code> getGrpcStatusCodes(Set<StatusCode> statusCodes) {
+    Set<Status.Code> returnCodes = new HashSet<>();
+    for (StatusCode code : statusCodes) {
```
It makes it easier to configure if all the status codes can be tossed in together.

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcTransportDescriptor.java:
```diff
@@ -0,0 +1,118 @@
+    exceptionContext.getResultFuture().setException(exceptionToThrow);
+  }
+
+  public ApiCallContext getDefaultCallContext() {
```
Ok, I will rename this.

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcTransportDescriptor.java:
```diff
@@ -0,0 +1,118 @@
+    return GrpcCallContext.getAsGrpcCallContextWithDefault(inputContext);
+  }
+
+  public ApiCallContext getCallContextWithTimeout(ApiCallContext callContext, Duration rpcTimeout) {
```
`HttpJsonAttemptCallable` was supposed to be deleted, so I will delete it. 

The issue is that while the code looks the same, the types are incompatible...

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingChannelProvider.java:
```diff
@@ -100,26 +86,42 @@ public boolean needsExecutor() {
-  public ManagedChannel getChannel(Executor executor) throws IOException {
-    if (!needsExecutor()) {
-      throw new IllegalStateException("getChannel(Executor) called when needsExecutor() is false");
+  public boolean needsHeaders() {
```
I guess it means "needs headers to be provided." If the `headerProvider` is not null, then we have what we need.

Both `needsExecutor()` and `needsHeaders()` have to be public because they tell the caller if they need to call `withExecutor` / `withHeaders` before calling `getTransportChannel()`.

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingChannelProvider.java:
```diff
@@ -100,26 +86,42 @@ public boolean needsExecutor() {
+    ScheduledExecutorService executor = executorProvider.getExecutor();
+    Map<String, String> headers = addGrpcVersion(headerProvider.getHeaders());
+
     List<ClientInterceptor> interceptors = Lists.newArrayList();
```
Oh! I never bothered to look at the javadoc :-P

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingChannelProvider.java:
```diff
@@ -128,76 +130,33 @@ private ManagedChannel createChannel(Executor executor) throws IOException {
+    for (Map.Entry<String, String> header : headers.entrySet()) {
+      if (header.getKey().equals(GoogleServiceHeaderProvider.getServiceHeaderLineKey())) {
+        newHeaders.put(
+            header.getKey(), header.getValue() + " grpc/" + GaxGrpcProperties.getGrpcVersion());
```
Yeah I think something along those lines would be cleaner. I'll probably add the ability to inject this logic into `GoogleServiceHeaderProvider` through the builder somehow.

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/longrunning/OperationsSettings.java:
```diff
@@ -129,9 +133,13 @@ public OperationsStub createStub() throws IOException {
+    return InstantiatingChannelProvider.newBuilder();
+  }
+
+  public static GoogleServiceHeaderProvider.Builder defaultGoogleServiceHeaderProviderBuilder() {
```
I will rename it to ApiClientHeaderProvider, and I will externalize the "x-goog-api-client" key.

garrettjonesgoogle commented on gax-httpjson/src/main/java/com/google/api/gax/httpjson/ApiResultRetryAlgorithm.java:
```diff
@@ -1,64 +0,0 @@
-/*
```
Only in the first pass when I copied all of the gax-grpc classes into gax-httpjson.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/ClientContext.java:
```diff
@@ -66,15 +68,18 @@
         .setBackgroundResources(Collections.<BackgroundResource>emptyList())
         .setExecutor(Executors.newScheduledThreadPool(0))
-        .setTransportContext(NullTransport.create())
+        .setHeaders(new HashMap<String, String>())
```
Will do.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/ExceptionContext.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.Set;
+
+@AutoValue
+public abstract class ExceptionContext {
```
But there are two other pieces of information needed also, the retryable codes and the result future. Essentially this is functioning like a request object for the `translateException` method in the `TransportDescriptor`.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/GoogleServiceHeaderProvider.java:
```diff
@@ -0,0 +1,186 @@
+public class GoogleServiceHeaderProvider implements HeaderProvider {
+  private static final String DEFAULT_VERSION = "";
+
+  public static String getServiceHeaderLineKey() {
```
Yeah, it was in `InstantiatingChannelProvider`. Because of changes I'm making due to other comments of yours, this is will no longer be a thing.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/GoogleServiceHeaderProvider.java:
```diff
@@ -0,0 +1,186 @@
+    headerLineParts.add(generatorName + "/" + generatorVersion);
+    headerLineParts.add("gax/" + GaxProperties.getGaxVersion());
+
+    return Joiner.on(" ").join(headerLineParts);
```
It's only called once. Although List + Joiner is heavier, it's expression is a little more elegant (spaces between elements don't need to be managed).

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/HeaderProvider.java:
```diff
@@ -0,0 +1,37 @@
+
+public interface HeaderProvider {
+
+  Map<String, String> getHeaders();
```
Currently this is only called once per client. This is just the provider; the usage of per-client, per-method, or per-request is a question of the calling sites.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/OperationCallableImpl.java:
```diff
@@ -0,0 +1,125 @@
+ *
+ * <p>Package-private for internal use.
+ */
+@InternalApi
```
I think that git didn't match it up because so many lines are different. I don't know of a way to make git match them up.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/OperationFuture.java:
```diff
@@ -52,12 +52,6 @@
   String getName() throws InterruptedException, ExecutionException;
 
   /**
-   * Returns the {@code OperationT} future of the initial request which started this {@code
```
It felt out of place to me... It was the one piece of the interface that requires the knowledge of how the operation is structured. Who needs the lower-level details that you're referring to? If we wanted to expose the underlying operation, it would have to be exposed as `OperationSnapshot`, which exposes the result and metadata as plain `Object` instead of `ResponseT` and `MetadataT`; if we want a typed result, we'd have to create another abstraction somehow.

vam-google replied:
Yes, this implies that each operation has a "starting operation" as its initialization. I think it is ok to have it on the surface. Without that there no way to get reasonable details about failed operation start. The metadata and response make sense only if the operation was started (i.e. initialized). This method provides the "initialization" status.

garrettjonesgoogle replied:
I will add this back, but returning `ApiFuture<OperationSnapshot>`.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/OperationSnapshot.java:
```diff
@@ -0,0 +1,54 @@
+  String getName();
+
+  /** Get metadata for the operation, which can change over the course of the operation. */
+  Object getMetadata();
```
Same response as elsewhere: I think this is lower-level than you were expecting; this represents the raw result from the service, and the response/metadata transformation is performed on top of it.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/OperationSnapshot.java:
```diff
@@ -0,0 +1,54 @@
+   * If the operation is done and successful, returns the response; if the operation is not done or
+   * if it is failed, returns null.
+   */
+  Object getResponse();
```
Same response as elsewhere: I think this is lower-level than you were expecting; this represents the raw result from the service, and the response/metadata transformation is performed on top of it.

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcOperationTransformers.java:
```diff
@@ -0,0 +1,126 @@
+import io.grpc.Status.Code;
+
+/** Package-private for internal use. */
+class GrpcOperationTransformers {
```
I feel like we can just keep this under `com.google.api.gax.grpc` even though core stuff is moving under `com.google.api.gax.longrunning`.

garrettjonesgoogle commented on gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpApiExceptionFactory.java:
```diff
@@ -63,6 +63,8 @@
   public static final String DATA_LOSS = "DATA_LOSS";
   public static final String UNKNOWN = "UNKNOWN";
 
+  public static final int STATUS_CODE_DEADLINE_EXCEEDED = 504;
```
I'll use the Apache http status code definition.

garrettjonesgoogle commented on gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpJsonTransportDescriptor.java:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  public ApiCallContext getCallContextWithTimeout(ApiCallContext callContext, Duration rpcTimeout) {
+    HttpJsonCallContext oldContext =
```
Not really, because of distinct types.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/ApiResultRetryAlgorithm.java:
```diff
@@ -27,15 +27,14 @@
 import org.threeten.bp.Duration;
 
 /* Package-private for internal use. */
+// FIXME after tests moved, make package-private again
```
Looks like I just didn't clean up this comment after multiple editing iterations.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/CallableFactory.java:
```diff
@@ -0,0 +1,323 @@
+    }
+  }
+
+  // FIXME after tests moved, make package-private again
```
They have already moved, so I will remove these FIXME's. 

The naming issue you call out applies to all of the methods in this class - should they all be renamed? (Last chance!)

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/AttemptCallable.java:
```diff
@@ -27,36 +27,37 @@
+      UnaryCallable<RequestT, ResponseT> callable,
+      RequestT request,
+      ApiCallContext callContext) {
+    this.transportDescriptor = Preconditions.checkNotNull(transportDescriptor);
```
I feel like it's easier to always require it if the class will explode in a subset of usage scenarios when it's not provided.

garrettjonesgoogle commented:
6. Header routing problem: my implementation is for static headers. I think we can make an additive change later for dynamic headers (which you would need). 
7. Look at the `withExecutor()` and `withHeaders()` methods on `TransportChannelProvider`, and see if it makes sense to you.

garrettjonesgoogle commented:
I have addressed feedback from @vam-google . PTAL if you want, or wait for a following update where I update docs & add tests.

garrettjonesgoogle commented:
@vam-google I have updated docs & tests, PTAL

garrettjonesgoogle commented:
I have created a diff PR for certain files as promised: https://github.com/googleapis/gax-java/pull/374

Note, it only include files that git didn't match up correctly; it doesn't include all files that were touched in the present PR.

garrettjonesgoogle commented:
@pongad I would recommend doing multiple passes on this PR and focus on different things in those passes; it would be hard to split this PR apart because it involves code moves & renames, so any splits would result in code that wouldn't compile.

One way to cut down on the lines to review is that you could completely skip gax-httpjson changes; they will mirror gax-grpc directly. Secondly, the most important thing to focus on is the surface under gax/ (transport-agnostic surfaces); prioritize reviewing that code. 

Lastly, the diff PR should help reviewing some files that git didn't match up for diffs.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/TransportDescriptor.java:
```diff
@@ -0,0 +1,72 @@
+        "TransportDescriptor.createDefaultCallContext not implemented");
+  }
+
+  public ApiCallContext getCallContextWithDefault(ApiCallContext inputContext) {
```
It was also to give a default if it wasn't set. This was an attempt to reduce code duplication. Maybe it doesn't make sense to bend the interface just for the purpose of reducing duplication.

pongad replied:
If it makes things easier, I don't have a great concern with it.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/rpc/NoHeaderProvider.java:
```diff
@@ -0,0 +1,42 @@
+
+  @Override
+  public Map<String, String> getHeaders() {
+    return new HashMap<>();
```
Sure.

garrettjonesgoogle commented:
@pongad I believe the callable stack was the right design. We can't really revisit the design if we want a third transport because we're going GA before such a thing will even be considered. Register your concerns now w.r.t. supporting a third transport!

pongad commented:
@garrettjonesgoogle I don't have specific concerns about adding new transports. Not having worked on gax-java for a while, I find it difficult to think of specific action items. I'm sorry I'm not being helpful, but here's my "nebulous" concern.

At the time of writing, gax-java master has 212 files and ~13,000 lines of code. At the same time, gax-go has 250 *statements*. I understand LOC isn't a good measure for quality or productivity, and to be fair, Java's job is harder (eg, Go can efficiently block; Java has to do things with executors). I don't think this should block GA, but there's something unsettling about the 52x size difference.

garrettjonesgoogle commented:
@vam-google , I have done the following:

* Added `initialFuture` back to `OperationFuture`
* Moved the `Fake*` classes to src/test

As I mentioned in the diff PR, there is a test for the transformers at GrpcOperationTransformersTest .

Is there anything else I'm missing?

I'd like to merge this to the gax-httpjson branch and then do another PR with additional changes before I merge to master.
